### ISSUE TITLE
property issues

### ISSUE THREAD

**Issue by hongzhidao** [nginx/njs#203]
@xeioex @drsm 

There are some issues happened in defineProperty. (see tests)

Here are the patches fix them.
https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4

My thoughts.
Fix issues first, then continue refactoring.

**Comment by hongzhidao** [comment]
@xeioex 

I'll do the refactor after the above issues, most of them are done.
One place I'm not sure. I want to rename all the `object` actually belong to `value` as `value`, `property` as `name`.
```
grep -r 'njs_value_t \*property' src/
src/njs_object_prop.c:    njs_value_t *value, njs_value_t *property)
src/njs_vmcode.c:    njs_value_t *object, njs_value_t *property, njs_value_t *retval);
...
src/njs_vmcode.c:njs_vmcode_property_in(njs_vm_t *vm, njs_value_t *object, njs_value_t *property)
src/njs_vmcode.c:    njs_value_t *property)
src/njs_value.h:    njs_value_t *object, njs_value_t *property);
src/njs_value.h:    njs_value_t *property, njs_value_t *retval);
src/njs_value.h:    njs_value_t *property, njs_value_t *value);
src/njs_object.h:    njs_value_t *value, njs_value_t *property);
src/njs_value.c:    const njs_value_t *property);
src/njs_value.c:    njs_value_t *property)
src/njs_value.c:    njs_object_t *object, const njs_value_t *property)
src/njs_value.c:njs_value_property(njs_vm_t *vm, njs_value_t *value, njs_value_t *property,
src/njs_value.c:    njs_value_t *property, njs_value_t *value)
``` 
What do you think?

BTW, I suggest cleaning the header files included in njs_clang.h and njs_value.h.
If OK, you help do this together with the above patches.

**Comment by drsm** [comment]
@hongzhidao 


> One place I'm not sure. I want to rename all the `object` actually belong to `value` as `value`, `property` as `name`.

Maybe better change `property` to `key`, as there also `Symbol` properties (not implemented yet, https://2ality.com/2014/12/es6-symbols.html#symbols-as-property-keys) .

**Comment by hongzhidao** [comment]
> Maybe better change property to key

Like this.
https://gist.github.com/hongzhidao/c6e43e1e83c3e4c83c88cac8f5c9a249

**Comment by xeioex** [comment]
@hongzhidao 

Agree, 'key' looks better.

**Comment by xeioex** [comment]
@hongzhidao 

> https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4#file-njs-propertys-patch-L406

Handling getters in njs_object_property() looks promising, but what about NJS_PROPERTY_HANDLER?

**Comment by hongzhidao** [comment]
> Why not directly use njs_value_property() in such places?

We do the fixes first, the codes append to njs_object_property() is from njs_value_property().
Later, there are changes with these functions.

My main ideas are:
1. There are two ways of accessing value property. Such as:
    delete a['b'];  // only query
    a['b']()  // get property value.
So njs_proerty_query() will still be public. And njs_value_property() is kept unchanged.
But getting property by string name is a helpful method. njs_object_property may be renamed as
njs_value_str_property().
You'll see there are some other APIs needed. For example.
njs_object_prop_get() is used in njs_object_prop_descriptor(). 
njs_object_prop_value() is used in njs_value_property() and njs_object_property().
All there will be shown in the following patches committed tomorrow.

> Handling getters in njs_object_property() looks promising, but what about NJS_PROPERTY_HANDLER?

I think NJS_PROPERTY_HANDLER does not need to be handled.
See https://github.com/nginx/njs/blob/master/src/njs_object_prop.c#L292

**Comment by xeioex** [comment]
@hongzhidao 

>  I want to rename all the object actually belong to value as value, property as name.

Looks useful. (property as key)

**Comment by hongzhidao** [comment]
@xeioex 
> I suggest cleaning the header files included in njs_clang.h and njs_value.h.

Are you on it?

> I want to rename all the object actually belong to value as value, property as name.

I'm on this.

**Comment by xeioex** [comment]
@hongzhidao 

> Are you on it?

will do it tomorrow.

**Comment by hongzhidao** [comment]
@xeioex 

Updated.
https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4

help tomorrow.
1. cleaning the header files included in njs_clang.h and njs_value.h.
2. simpilify `prop->value.data.u.prop_handler(vm, value, setv...)`.

**Comment by hongzhidao** [comment]
@xeioex 

njs.dump() and JSON.stringify issues. 
```
>> var o = {}; Object.defineProperty(o, 'a', { value: 'bar' })
{
}
>> >> njs.dump(o)
'{}'
>> JSON.stringify(o)
'{}'
```
I can't find the definite places, help fix them.

**Comment by drsm** [comment]
@hongzhidao 

> njs.dump() and JSON.stringify issues.
> 
> ```
> >> var o = {}; Object.defineProperty(o, 'a', { value: 'bar' })
> {
> }
> >> >> njs.dump(o)
> '{}'
> >> JSON.stringify(o)
> '{}'
> ```

It's OK, as we're dumping only enumerable properties.
```js
>> var o = Object.defineProperty({}, 'a', { value: 'bar', configurable: true })
undefined
>> njs.dump(o)
'{}'
>> var o = Object.defineProperty(o, 'a', { enumerable: true })
undefined
>> njs.dump(o)
'{a:'bar'}'
```
the problem:
```js
>> var o = Object.defineProperty({}, 'a', { get: () => 1, enumerable: true })
undefined
>> JSON.stringify(o)
'{}'
>> njs.dump(o)
'{a:<empty>}'
```
`stringify` sholud invoke an accessor, `dump` should show what accessors are there:
```js
> var o = Object.defineProperty({}, 'a', { get: () => 1, enumerable: true })
undefined
> JSON.stringify(o)
'{"a":1}'
> o
{ a: [Getter] }
```

**Comment by hongzhidao** [comment]
@xeioex 

> stringify sholud invoke an accessor

njs_object_prop_value() can help.

**Comment by hongzhidao** [comment]
@drsm welcome to test.
 
https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4
Fixed Error().
Fixed njs_vmcode_method_frame().
Fixed njs_object_property().
Added getter/setter literal support.

what's left: fix dump() issue.

**Comment by drsm** [comment]
@hongzhidao 
please update:
```diff
diff -r 68c97f3a6a2a src/njs_object_prop.c
--- a/src/njs_object_prop.c	Mon Aug 05 01:32:54 2019 -0400
+++ b/src/njs_object_prop.c	Mon Aug 05 08:48:27 2019 +0300
@@ -329,7 +329,6 @@
     njs_int_t           ret;
     njs_bool_t          data, accessor;
     njs_value_t         value;
-    njs_lvlhsh_query_t  lhq;
 
     data = 0;
     accessor = 0;
@@ -378,9 +377,6 @@
         return ret;
     }
 
-    lhq.key = njs_str_value("value");
-    lhq.key_hash = NJS_VALUE_HASH;
-
     ret = njs_value_str_property(vm, desc, &njs_str_value("value"),
                                  NJS_VALUE_HASH, &value);
 
@@ -394,9 +390,6 @@
         return ret;
     }
 
-    lhq.key = njs_str_value("writable");
-    lhq.key_hash = NJS_WRITABABLE_HASH;
-
     ret = njs_value_str_property(vm, desc, &njs_str_value("writable"), 
                                  NJS_WRITABABLE_HASH, &value);
 
@@ -410,9 +403,6 @@
         return ret;
     }
 
-    lhq.key = njs_str_value("enumerable");
-    lhq.key_hash = NJS_ENUMERABLE_HASH;
-
     ret = njs_value_str_property(vm, desc, &njs_str_value("enumerable"),
                                  NJS_ENUMERABLE_HASH, &value);
 
@@ -425,9 +415,6 @@
         return ret;
     }
 
-    lhq.key = njs_str_value("configurable");
-    lhq.key_hash = NJS_CONFIGURABLE_HASH;
-
     ret = njs_value_str_property(vm, desc, &njs_str_value("configurable"),
                                  NJS_CONFIGURABLE_HASH, &value);
 
diff -r 68c97f3a6a2a src/njs_parser_terminal.c
--- a/src/njs_parser_terminal.c	Mon Aug 05 01:32:54 2019 -0400
+++ b/src/njs_parser_terminal.c	Mon Aug 05 08:48:27 2019 +0300
@@ -544,7 +544,7 @@
 static njs_token_t
 njs_parser_object_token(njs_vm_t *vm, njs_parser_t *parser)
 {
-    size_t       offset;
+    size_t       offset = 0;
     njs_str_t    *name;
     njs_token_t  token;
 
@@ -593,7 +593,7 @@
 njs_parser_object_property(njs_vm_t *vm, njs_parser_t *parser,
     njs_lexer_t *lexer, njs_parser_node_t *obj, njs_token_t token)
 {
-    uint32_t               hash, token_line;
+    uint32_t               hash = 0, token_line = 0;
     njs_int_t              ret;
     njs_str_t              name;
     njs_token_t            prop_token;
diff -r 68c97f3a6a2a src/njs_value.c
--- a/src/njs_value.c	Mon Aug 05 01:32:54 2019 -0400
+++ b/src/njs_value.c	Mon Aug 05 08:48:27 2019 +0300
@@ -122,7 +122,6 @@
     njs_int_t           ret;
     njs_uint_t          tries;
     njs_value_t         fun, retval;
-    njs_lvlhsh_query_t  lhq;
 
     static const uint32_t  hashes[] = {
         NJS_VALUE_OF_HASH,
@@ -149,9 +148,6 @@
         if (njs_is_object(value) && tries < 2) {
             hint ^= tries++;
 
-            lhq.key_hash = hashes[hint];
-            lhq.key = names[hint];
-
             ret = njs_value_str_property(vm, value, (njs_str_t *) &names[hint],
                                          hashes[hint], &fun);
 
```

**Comment by drsm** [comment]
@hongzhidao 

this still fails:
```js
//node
> var get = 'get', o = { get }
undefined
> o
{ get: 'get' }
> var set = 'set', o = { set }
undefined
> o
{ set: 'set' }
```
see https://github.com/nginx/njs/issues/118#issuecomment-516453211il

**Comment by xeioex** [comment]
@hongzhidao 

> https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4
> Fixed njs_object_property().
> Refactored out njs_value_str_property().

I think these two patches should be combined into one patch (after "Style" patch).

IMO, it a bad idea to change a function which will be removed later (in the next patch).

I am on merging 'Style' patch.

**Comment by xeioex** [comment]
@hongzhidao 

>  njs_int_t njs_value_property(njs_vm_t *vm, njs_value_t *value, njs_value_t *key, njs_value_t *retval);
> njs_int_t njs_value_str_property(njs_vm_t *vm, const njs_value_t *value, njs_str_t *key, uint32_t key_hash, njs_value_t *retval);

I think njs_value_property() should convert property if necessary and call njs_value_str_property().

**Comment by hongzhidao** [comment]
@xeioex 

> I think njs_value_property() should convert key if necessary and call njs_value_str_property().
Otherwise the duplicated each other.

Notice that `njs_object_proerty()` (now is `njs_value_str_property()`) is called under the condition `njs_is_object().`

**Comment by xeioex** [comment]
@hongzhidao 

> Notice that njs_object_property() (now is njs_value_str_property()) is only called under the condition njs_is_object().

I would like to have njs_value_str_property() as optimized version of njs_value_property() when key is known in advance (many places).

The main concern for me with current `njs_value_str_property()` is that it does  not handle NJS_PROPERTY_HANDLER. While I see that currently NJS_PROPERTY_HANDLER is used only for "length", "__proto__", "prototype" and "constructor" and currently it is not possible to get NJS_PROPERTY_HANDLER in  `njs_value_str_property()` call.

**Comment by xeioex** [comment]
@hongzhidao 

I suggest to eliminate NJS_METHOD first.

**Comment by hongzhidao** [comment]
> @hongzhidao
> 
> I suggest to eliminate NJS_METHOD first.

I thought about it. It seems we can distinguish method and property by is_function if eliminate NJS_METHOD. But all the benefits of eliminating NJS_METHOD are limited. So I kept it unchanged.

I suggest we do it later.

**Comment by xeioex** [comment]
@hongzhidao 

> The most important thing is to fix current issues, such as njs_object_property(), dump() and stringify().

Agree. But I think njs_value_str_property() vs njs_value_property() can be improved.

I am on it.

**Comment by hongzhidao** [comment]
OK, continue tomorrow.

**Comment by hongzhidao** [comment]
@drsm 

What specification of JS language would you like to suggest?
Now, I usually refer to this. https://tc39.es/ecma262/

**Comment by drsm** [comment]
@hongzhidao 

I'm also suggest to refer to [the latest draft](https://tc39.es/ecma262/) and [proposals](https://github.com/tc39/proposals) like [BigInt](https://github.com/tc39/proposal-bigint)

**Comment by xeioex** [comment]
@hongzhidao 

At this point I do not see the crystal clear path on how to eliminate njs_object_property() in good way.

I still want to merge it with njs_value_property(). At this moment, I am on simplifying njs_property_query(). API at this moment is too difficult.
What can be improved:
1) `pq.shared` can be removed. https://gist.github.com/a4ac11e8bb45db6027349491373f2216 (in progress).
2) `NJS_METHOD` type can be removed.
..

**Comment by hongzhidao** [comment]
OK, tell me if there is anyting I can help.

**Comment by xeioex** [comment]
@hongzhidao 

You can review these two patches.

**Comment by xeioex** [comment]
@hongzhidao 

BTW, I think NJS_PROPERTY_HANDLER can also be eliminated.
converted to NJS_PROPERTY with function value (similar to NJS_METHOD).

This would allow us to to have your version of njs_object_property() which would handle all cases and be generic.

**Comment by hongzhidao** [comment]
It'll be very good if they are all removed. Wait for your patch :)
I'm on reivew now.

**Comment by hongzhidao** [comment]
>  I think NJS_PROPERTY_HANDLER can also be eliminated.

Can you also consider one place.
```
typedef struct {
    njs_lvlhsh_query_t          lhq;

    /* scratch is used to get the value of an NJS_PROPERTY_HANDLER property. */
    njs_object_prop_t           scratch;   // can we remove it? Just an idea.
...
```

**Comment by xeioex** [comment]
@hongzhidao 

Updated.
https://gist.github.com/0268b795b772a3d073029ea5372b0e1b

**Comment by hongzhidao** [comment]
@xeioex 

> pq.shared can be removed. https://gist.github.com/a4ac11e8bb45db6027349491373f2216

Looks good. But put this line in front. https://github.com/nginx/njs/blob/master/src/njs_object_prop.c#L648

**Comment by xeioex** [comment]
@hongzhidao 

NJS_PROPERTY_HANDLER is also copied by this function.

prop->value is invalid for it. But we need to copy it too, otherwise "delete" operation would SEGFAULT on writing to it (read only memory).

**Comment by hongzhidao** [comment]
@xeioex 

The key point is that NJS_PROPERTY_HANDLER, NJS_METHOD and NJS_PROPERTY are copied in shared mode and copy function if is function. Right?

**Comment by xeioex** [comment]
@hongzhidao 

> The key point is that NJS_PROPERTY_HANDLER, NJS_METHOD and NJS_PROPERTY are copied in shared mode (found in shared_hash) and copy function if is function. Right?

yes.

**Comment by hongzhidao** [comment]
> I think NJS_PROPERTY_HANDLER can also be eliminated.

Are you on it? 

It seems it'll make a big change.

**Comment by xeioex** [comment]
> It seems it'll make a big change.

yes. I am on rethinking what NJS_PROPERTY_HANDLER is actually for.

It is not the same as ordinary getter.

**Comment by xeioex** [comment]
@hongzhidao 

My ideas:
1) rename NJS_PROPERTY_HANDLER to NJS_PROPERTY_ON_DEMAND
For properties like "prototype", "constructor" they should be called once at the first access.
NJS_PROPERTY_ON_DEMAND will be called inside njs_property_query() and hidden from top level functions.

2) for njs_array_length  we need a special handling.
3) njs_function_length will be created in njs_prop_private_copy() (the same way as "name" now).

**Comment by hongzhidao** [comment]
@xeioex 

See this first. https://raw.githubusercontent.com/ldarren/QuickJS/mod/quickjs.c  (find "__proto__")
I'm also not sure it's OK to regard NJS_PROPERTY_HANDLER as getter.
But it can make things easier. NJS_PROPERTY_HANDLER can be easily removed, and the codes are concise.
It's worth to rethink what the adverse effect is if regard them as getter.

What's your suggestion? @drsm 
We want to make them as getter, like [].length, __proto__, etc.

**Comment by xeioex** [comment]
@hongzhidao 

> I'm also not sure it's OK to regard NJS_PROPERTY_HANDLER as getter.

Definitely not. `Object.getOwnPropertyDescriptor(Object, 'prototype')` should return data descriptor.
See NJS_PROPERTY_ON_DEMAND in https://github.com/nginx/njs/issues/203#issuecomment-518753753.

**Comment by xeioex** [comment]
@hongzhidao 

I think the first step will be elimination of NJS_PROPERTY_HANDLER in NJS_PROPERTY_QUERY_GET query mode. Complete removal is harder. But, for `njs_object_property()` it is enough.

**Comment by hongzhidao** [comment]
> I think the first step will be elimination of NJS_PROPERTY_HANDLER in NJS_PROPERTY_QUERY_GET query mode. Complete removal is harder. But, for njs_object_property() it is enough.

OK.

**Comment by xeioex** [comment]
@hongzhidao 

> BTW, it's OK to keep NJS_PROPERTY_HANDLER, it's clear in NJS.

Agree. I still want to improve  NJS_PROPERTY_HANDLER but I think this can be done in the next refactoring (good refactoring requires many changes).

> Fixed njs_object_property().
> Added getter/setter literal support.

You may want to update the patch. I think `njs_object_property()` name is OK for now (no need to rename it as njs_value_str_property()).
I plan to review it tomorrow.

**Comment by hongzhidao** [comment]
@xeioex 

Fix njs_object_property() first, then consider continuning refactor or updating getter/setter patch.
https://gist.github.com/hongzhidao/4f0dbc8640804a47727444ff4744a34e

**Comment by xeioex** [comment]
@hongzhidao 

https://gist.github.com/hongzhidao/4f0dbc8640804a47727444ff4744a34e#file-njs-0807-patch-L478

please, do not remove these 'case NJS_ERROR:'
1) it adds clarity
2) it is a seraparate change.

Otherwise looks good, will commit with some improvements and fixes.
1) njs_object_property() should set undefined value to retval in case of NJS_DECLINED
2) introduced njs_object_property_init() to set lhq->proto only once.
3) avoid using switch statement when only 3 possible return values (NJS_ERROR usually is treated separately).
4) more tests

https://gist.github.com/fc3ae94b77d4a210bf35c686379f351d

**Comment by hongzhidao** [comment]
@xeioex 

Forgot to use njs_object_property_init()?
https://gist.github.com/xeioex/fc3ae94b77d4a210bf35c686379f351d#file-hd-object-prop-patch-L340

**Comment by xeioex** [comment]
@hongzhidao 

> Forgot to use njs_object_property_init()?

No, because it is only init (sets lhq->proto).

> Make this as a separate patch?

no, `case NJS_ERROR:` here is for a reason.

**Comment by hongzhidao** [comment]
No problem. Then what do you plan to do after this?
https://gist.github.com/fc3ae94b77d4a210bf35c686379f351d

**Comment by xeioex** [comment]
I want to finish #40 first before #118.

#40 fixes 3 more V8 benchmarks.

**Comment by hongzhidao** [comment]
There are still bugs with njs.dump() and JSON.stringify()
https://github.com/nginx/njs/issues/203#issuecomment-518084627
It needs your help :).
Getter/setter patch depends on it.

**Comment by xeioex** [comment]
@hongzhidao 

> But I also can commit getter/setter patch first, what's your suggestion?

Yes, it looks like a separate issue.

**Comment by xeioex** [comment]
@hongzhidao 

Added support for accessor property descriptors in njs.dump().
https://gist.github.com/c890e75bac330e64c22bd56952d2736e

Please review.

**Comment by hongzhidao** [comment]
> https://gist.github.com/c890e75bac330e64c22bd56952d2736e

Looks good. One tiny detail: string_getter_setter vs string_getset?

**Comment by xeioex** [comment]
@hongzhidao  @drsm 

> https://gist.github.com/cc5b059a9449ae8a076575f086bfae40

Added support for accessor property descriptors in njs.dump().
Added support for accessor property descriptors in JSON.stringify().

**Comment by hongzhidao** [comment]
review it tomorrow, and I’ll update getter/setter patch.

**Comment by hongzhidao** [comment]
> https://gist.github.com/660eaedbd6b0b85ac5a14b4c71da6476

Looks good.

**Comment by hongzhidao** [comment]
@drsm @xeioex 

Added getter/setter literal support.
https://gist.github.com/hongzhidao/1c4bf3f02ffbdd605e6252a7f48373b4

Welcome to test.
